expo/expo
# [expo] Fix deactivating keep awake on Android when closing

# Why

Fixes #12584

# How

After debugging the unhandled promise I noticed this error originates when closing the app (with the triangle, as described in the issue). This is caused by the activity manager not having a `currentActivity` anymore and thus would fail to release the keep awake lock. The unhandled promise exception does show after you reopen the app.

This fix avoids calling `deactivateKeepAwake` on Android when the app is closing. On iOS, it still does and web has its own ExpoRoot without keep awake locks.

> Using `useKeepAwake` in a managed project with Expo Go doesn't trigger this behavior. BUT in a bare project, it also triggers this method when used in the app code.

### Android

Looking at the [Android documentation](https://developer.android.com/training/scheduling/wakelock#java) I found:

<img width="878" alt="Screenshot 2021-06-29 at 14 43 57" src="https://user-images.githubusercontent.com/1203991/123799376-744abf00-d8e8-11eb-956d-b76c0f169a88.png">

This would mean that we don't have to release the lock on Android. It's handled by the OS when using `FLAG_KEEP_SCREEN_ON`. We don't even need the `WAKE_LOCK` permission 🤔 


### iOS

On iOS [this is different](https://developer.apple.com/documentation/uikit/uiapplication/1623070-idletimerdisabled), we actually have to release the lock once we are done. I'm not sure if the error from #12584 pops up for iOS, I don't think it does.

<img width="769" alt="Screenshot 2021-06-29 at 14 46 36" src="https://user-images.githubusercontent.com/1203991/123799759-d3103880-d8e8-11eb-90b3-8097a8454895.png">


# Alternatives

There are a few other options that we can explore.

1. Return the promise from `expo-keep-awake` to "hide" the exception on Android.
2. Avoiding `expo-keep-awake` in the Expo Go client, using `android:keepScreenOn="true"` in the `AndroidManifest.xml`


# Test Plan

- Create a new project
- Open in emulator _(I'm using a Pixel 3 with API 29 (Q))_
- After app is loaded, press back (triangle button). _This throws the native exception_
- Press active apps (square button)
- Open the app again _This pops up the previously thrown exception as "unhandled promise"_

After this fix, the unhandled promise exception should be gone.

- [Here is a patch for the fix](https://github.com/expo/expo/files/6733504/expo%2B42.0.0-beta.2.patch.txt), usable with [patch-package](https://github.com/ds300/patch-package). Rename `.patch.txt` to `.patch`.


# Checklist
